common/src/main/java/com/microsoft/identity/common/internal/ui/embeddedwebview/challengehandlers/PKeyAuthChallengeHandler.java

```diff
@@ -22,21 +22,259 @@
+import android.support.annotation.NonNull;
+import android.webkit.WebView;
 
-import com.microsoft.identity.common.internal.providers.oauth2.AuthorizationResponse;
+import com.microsoft.identity.common.adal.internal.AuthenticationConstants;
+import com.microsoft.identity.common.adal.internal.AuthenticationSettings;
+import com.microsoft.identity.common.adal.internal.IDeviceCertificate;
+import com.microsoft.identity.common.adal.internal.JWSBuilder;
+import com.microsoft.identity.common.adal.internal.util.StringExtensions;
+import com.microsoft.identity.common.exception.ClientException;
+import com.microsoft.identity.common.exception.ErrorStrings;
+import com.microsoft.identity.common.internal.logging.Logger;
+import com.microsoft.identity.common.internal.providers.oauth2.AuthorizationRequest;
 
-final class PKeyAuthChallengeHandler {
-    private Handler mRefHandler; //NOPMD
+import java.lang.reflect.Constructor;
+import java.lang.reflect.InvocationTargetException;
+import java.security.interfaces.RSAPrivateKey;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
 
-    private AuthorizationResponse mResponse; //NOPMD
+import static com.microsoft.identity.common.exception.ErrorStrings.DEVICE_CERTIFICATE_REQUEST_INVALID;
 
-    PKeyAuthChallengeHandler(Handler ref, AuthorizationResponse response) {
-        mRefHandler = ref;
-        mResponse = response;
+public final class PKeyAuthChallengeHandler implements IChallengeHandler {
+    private static final String TAG = PKeyAuthChallengeHandler.class.getSimpleName();
+    /**
+     * Certificate authorities are passed with delimiter.
+     */
+    private static final String CHALLENGE_REQUEST_CERT_AUTH_DELIMITER = ";";
+    private WebView mWebView;
+    private String mRedirectUri;
+    private AuthorizationRequest mRequest;
+    private IChallengeCompletionCallback mChallengeCallback;
+
+    /**
+     * @param redirectUri        Location: urn:http-auth:CertAuth?Nonce=<noncevalue>
+     *                           &CertAuthorities=<distinguished names of CAs>&Version=1.0
+     *                           &SubmitUrl=<URL to submit response>&Context=<server state that
+     *                           client must convey back>
+     * @param view
+     * @param completionCallback
+     * @throws ClientException
+     */
+    public PKeyAuthChallengeHandler(@NonNull final String redirectUri,
+                                    @NonNull final WebView view,
+                                    @NonNull final AuthorizationRequest request,
+                                    @NonNull IChallengeCompletionCallback completionCallback) {
+        mWebView = view;
+        mRedirectUri = redirectUri;
+        mRequest = request;
+        mChallengeCallback = completionCallback;
+    }
+
+    public void process() {
+        mWebView.stopLoading();
+        mChallengeCallback.setPKeyAuthStatus(true);
+        new Thread(new Runnable() {

```

Comment:
since you're posting a Runnable to the MessageQueue in the UI thread, you don't need to start a background thread to post the Runnable.